### PR TITLE
Add auto-generated doc to Swift4 generator

### PR DESCRIPTION
**Pull request by wing328**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Add auto-generated documentation (README, API doc, model doc) to Swift4 generator

@jgavris (2017/07) @ehyche (2017/08) @Edubits (2017/09) @jaz-ah (2017/09) @d-date (2018/03)

### CODE REVIEW

File: samples/client/petstore/swift4/default/docs/StoreAPI.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+```
+
+Place an order for a pet
+
+### Example 
+```swift
+// the following code samples are still beta. If you notice any issue, please report via http://github.com/OpenAPITools/openapi-generator/issues/new
+import PetstoreClient
+
+let body = Order(123, 123, 123, "TODO", "status_example", 123) // Order | order placed for purchasing the pet
```
Comment (d-date) [inline samples/client/petstore/swift4/default/docs/StoreAPI.md]: - The initializer should have labels for parameters. 
- It looks `Bool` turns `Int`, should be `true` or `false`.

As is:
```swift
let body = Order(123, 123, 123, "TODO", "status_example", 123)
```
Ideal:
```swift
let body = Order(id: 123, petId: 123, quantity: 123, shipDate: "TODO", status: "status_example", complete: true)
```

File: modules/openapi-generator/src/main/resources/swift4/api_doc.mustache [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                // when the promise is fulfilled
+            }.always {
+                // regardless of whether the promise is fulfilled, or rejected
+            }.catch { errorType in
+                // when the promise is rejected
+        }
+}
+{{/usePromiseKit}}
+{{#useRxSwift}}
+// TODO rxswift sample code. To contribute, please open a ticket via http://github.com/OpenAPITools/openapi-generator/issues/new
```
Comment (d-date) [inline modules/openapi-generator/src/main/resources/swift4/api_doc.mustache]: If indent is not match, please indent with 4 spaces

```suggestion
{{#summary}}// {{{.}}}
{{/summary}}{{classname}}.{{{operationId}}}({{#allParams}}{{paramName}}: {{paramName}}{{#hasMore}}, {{/hasMore}}{{/allParams}}).subscribe(onNext: {
        // when request is on success
    }, onError: { error in
        // when request was failed with error
    }).disposed(by: disposeBag)
}
```

File: samples/client/petstore/swift4/default/docs/FormatTest.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+Name | Type | Description | Notes
+------------ | ------------- | ------------- | -------------
+**integer** | **Int** |  | [optional] 
+**int32** | **Int** |  | [optional] 
+**int64** | **Int64** |  | [optional] 
+**number** | **Double** |  | 
+**float** | **Float** |  | [optional] 
+**double** | **Double** |  | [optional] 
+**string** | **String** |  | [optional] 
+**byte** | [**Data**](Data.md) |  |
```
Comment (d-date) [inline samples/client/petstore/swift4/default/docs/FormatTest.md]: Data is standard struct in Swift.

File: samples/client/petstore/swift4/default/docs/FormatTest.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+------------ | ------------- | ------------- | -------------
+**integer** | **Int** |  | [optional] 
+**int32** | **Int** |  | [optional] 
+**int64** | **Int64** |  | [optional] 
+**number** | **Double** |  | 
+**float** | **Float** |  | [optional] 
+**double** | **Double** |  | [optional] 
+**string** | **String** |  | [optional] 
+**byte** | [**Data**](Data.md) |  | 
+**binary** | [**URL**](URL.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/default/docs/FormatTest.md]: URL is standard struct in Swift

File: samples/client/petstore/swift4/default/docs/FormatTest.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+**integer** | **Int** |  | [optional] 
+**int32** | **Int** |  | [optional] 
+**int64** | **Int64** |  | [optional] 
+**number** | **Double** |  | 
+**float** | **Float** |  | [optional] 
+**double** | **Double** |  | [optional] 
+**string** | **String** |  | [optional] 
+**byte** | [**Data**](Data.md) |  | 
+**binary** | [**URL**](URL.md) |  | [optional] 
+**date** | [**Date**](Date.md) |  |
```
Comment (d-date) [inline samples/client/petstore/swift4/default/docs/FormatTest.md]: Date is standard struct in Swift

File: samples/client/petstore/swift4/default/docs/FormatTest.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+**int32** | **Int** |  | [optional] 
+**int64** | **Int64** |  | [optional] 
+**number** | **Double** |  | 
+**float** | **Float** |  | [optional] 
+**double** | **Double** |  | [optional] 
+**string** | **String** |  | [optional] 
+**byte** | [**Data**](Data.md) |  | 
+**binary** | [**URL**](URL.md) |  | [optional] 
+**date** | [**Date**](Date.md) |  | 
+**dateTime** | [**Date**](Date.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/default/docs/FormatTest.md]: Same as above

File: samples/client/petstore/swift4/default/docs/FormatTest.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+**int64** | **Int64** |  | [optional] 
+**number** | **Double** |  | 
+**float** | **Float** |  | [optional] 
+**double** | **Double** |  | [optional] 
+**string** | **String** |  | [optional] 
+**byte** | [**Data**](Data.md) |  | 
+**binary** | [**URL**](URL.md) |  | [optional] 
+**date** | [**Date**](Date.md) |  | 
+**dateTime** | [**Date**](Date.md) |  | [optional] 
+**uuid** | [**UUID**](UUID.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/default/docs/FormatTest.md]: UUID is struct in Foundation, no need to reference.

File: samples/client/petstore/swift4/objcCompatible/docs/AdditionalPropertiesClass.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+------------ | ------------- | ------------- | -------------
+**mapString** | **[String:String]** |  | [optional] 
+**mapNumber** | **[String:Double]** |  | [optional] 
+**mapInteger** | **[String:Int]** |  | [optional] 
+**mapBoolean** | **[String:Bool]** |  | [optional] 
+**mapArrayInteger** | [**[String:[Int]]**](Array.md) |  | [optional] 
+**mapArrayAnytype** | [**[String:[Any]]**](Array.md) |  | [optional] 
+**mapMapString** | [**[String:[String:String]]**](Dictionary.md) |  | [optional] 
+**mapMapAnytype** | [**[String:[String:Any]]**](Dictionary.md) |  | [optional] 
+**anytype1** | [**Any**](.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/objcCompatible/docs/AdditionalPropertiesClass.md]: Any is primitive type in Swift

File: samples/client/petstore/swift4/objcCompatible/docs/ArrayTest.md [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# ArrayTest
+
+## Properties
+Name | Type | Description | Notes
+------------ | ------------- | ------------- | -------------
+**arrayOfString** | **[String]** |  | [optional] 
+**arrayArrayOfInteger** | [**[[Int64]]**](Array.md) |  | [optional] 
+**arrayArrayOfModel** | [**[[ReadOnlyFirst]]**](Array.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/objcCompatible/docs/ArrayTest.md]: should not refer to Array.md, but ReadOnlyFirst.md.

File: samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+[[Back to top]](#) [[Back to API list]](../README.md#documentation-for-api-endpoints) [[Back to Model list]](../README.md#documentation-for-models) [[Back to README]](../README.md)
+
+# **testEndpointParameters**
+```swift
+    open class func testEndpointParameters(number: Double, double: Double, patternWithoutDelimiter: String, byte: Data, integer: Int? = nil, int32: Int? = nil, int64: Int64? = nil, float: Float? = nil, string: String? = nil, binary: URL? = nil, date: Date? = nil, dateTime: Date? = nil, password: String? = nil, callback: String? = nil, completion: @escaping ((_ data: Void?,_ error: Error?) -> Void))
+```
+
+Fake endpoint for testing various parameters 假端點 偽のエンドポイント 가짜 엔드 포인트 
+
+Fake endpoint for testing various parameters 假端點 偽のエンドポイント 가짜 엔드 포인트
```
Comment (d-date) [inline samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md]: Is it correct to be printed out same comments?

File: samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+ **number** | **Double**| None | 
+ **double** | **Double**| None | 
+ **patternWithoutDelimiter** | **String**| None | 
+ **byte** | **Data**| None | 
+ **integer** | **Int**| None | [optional] 
+ **int32** | **Int**| None | [optional] 
+ **int64** | **Int64**| None | [optional] 
+ **float** | **Float**| None | [optional] 
+ **string** | **String**| None | [optional] 
+ **binary** | **URL****URL**| None | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md]: **URL****URL** should be **URL**

File: samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+        dump(response)
+    }
+}
+```
+
+### Parameters
+
+Name | Type | Description  | Notes
+------------- | ------------- | ------------- | -------------
+ **param** | [**[String:String]**](String.md)| request body |
```
Comment (d-date) [inline samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md]: String.md should not be refer.

File: samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+    }
+}
+```
+
+### Parameters
+
+Name | Type | Description  | Notes
+------------- | ------------- | ------------- | -------------
+ **param** | **String**| field1 | 
+ **param2** | **String**| field2 |
```
Comment (d-date) [inline samples/client/petstore/swift4/objcCompatible/docs/FakeAPI.md]: [NIT] should have space after `**String**`

File: samples/client/petstore/swift4/promisekit/docs/AdditionalPropertiesClass.md [hunk 14]
```diff
@@ -0,0 +1,11 @@
+# AdditionalPropertiesClass
+
+## Properties
+Name | Type | Description | Notes
+------------ | ------------- | ------------- | -------------
+**mapString** | **[String:String]** |  | [optional] 
+**mapMapString** | [**[String:[String:String]]**](Dictionary.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/AdditionalPropertiesClass.md]: Should not refer Dictionary.md

File: samples/client/petstore/swift4/promisekit/docs/ArrayOfArrayOfNumberOnly.md [hunk 15]
```diff
@@ -0,0 +1,10 @@
+# ArrayOfArrayOfNumberOnly
+
+## Properties
+Name | Type | Description | Notes
+------------ | ------------- | ------------- | -------------
+**arrayArrayNumber** | [**[[Double]]**](Array.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/ArrayOfArrayOfNumberOnly.md]: Should not refer Array.md

File: samples/client/petstore/swift4/promisekit/docs/ArrayTest.md [hunk 16]
```diff
@@ -0,0 +1,12 @@
+# ArrayTest
+
+## Properties
+Name | Type | Description | Notes
+------------ | ------------- | ------------- | -------------
+**arrayOfString** | **[String]** |  | [optional] 
+**arrayArrayOfInteger** | [**[[Int64]]**](Array.md) |  | [optional] 
+**arrayArrayOfModel** | [**[[ReadOnlyFirst]]**](Array.md) |  | [optional]
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/ArrayTest.md]: Array.md

File: samples/client/petstore/swift4/promisekit/docs/FakeAPI.md [hunk 17]
```diff
@@ -0,0 +1,638 @@
...
+[**testEndpointParameters**](FakeAPI.md#testendpointparameters) | **POST** /fake | Fake endpoint for testing various parameters 假端點 偽のエンドポイント 가짜 엔드 포인트 
+[**testEnumParameters**](FakeAPI.md#testenumparameters) | **GET** /fake | To test enum parameters
+[**testGroupParameters**](FakeAPI.md#testgroupparameters) | **DELETE** /fake | Fake endpoint to test group parameters (optional)
+[**testInlineAdditionalProperties**](FakeAPI.md#testinlineadditionalproperties) | **POST** /fake/inline-additionalProperties | test inline additionalProperties
+[**testJsonFormData**](FakeAPI.md#testjsonformdata) | **GET** /fake/jsonFormData | test json serialization of form data
+
+
+# **fakeOuterBooleanSerialize**
+```swift
+    open class func fakeOuterBooleanSerialize( body: Bool? = nil) -> Promise<Bool>
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/FakeAPI.md]: [NIT] should be removed space before parameter

File: samples/client/petstore/swift4/promisekit/docs/FakeAPI.md [hunk 18]
```diff
@@ -0,0 +1,638 @@
...
+let number = 987 // Double | None
+let double = 987 // Double | None
+let patternWithoutDelimiter = "patternWithoutDelimiter_example" // String | None
+let byte = TODO // Data | None
+let integer = 987 // Int | None (optional)
+let int32 = 987 // Int | None (optional)
+let int64 = 987 // Int64 | None (optional)
+let float = 987 // Float | None (optional)
+let string = "string_example" // String | None (optional)
+let binary = "TODO" // URL | None (optional)
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/FakeAPI.md]: URL example is like below. String should be more exampled.

```swift
let binary = URL(string: "https://example.com")
```

File: samples/client/petstore/swift4/promisekit/docs/FakeAPI.md [hunk 19]
```diff
@@ -0,0 +1,638 @@
...
+let date = TODO // Date | None (optional)
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/FakeAPI.md]: Date example is like below.

```swift
let date = Date() // means now.
```

File: samples/client/petstore/swift4/promisekit/docs/FakeAPI.md [hunk 20]
```diff
@@ -0,0 +1,638 @@
...
+let dateTime = TODO // Date | None (optional)
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/FakeAPI.md]: Datetime is also.

```swift
let datetime = Date() // means now.
```

File: samples/client/petstore/swift4/promisekit/docs/FakeAPI.md [hunk 21]
```diff
@@ -0,0 +1,638 @@
...
+                // when the promise is rejected
+        }
+}
+```
+
+### Parameters
+
+Name | Type | Description  | Notes
+------------- | ------------- | ------------- | -------------
+ **param** | [**[String:String]**](String.md)| request body |
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/FakeAPI.md]: String.md)

File: samples/client/petstore/swift4/promisekit/docs/FakeAPI.md [hunk 22]
```diff
@@ -0,0 +1,638 @@
...
+### Parameters
+
+Name | Type | Description  | Notes
+------------- | ------------- | ------------- | -------------
+ **param** | **String**| field1 | 
+ **param2** | **String**| field2 | 
+
+### Return type
+
+void (empty response body)
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/FakeAPI.md]: type name void should be capitalized `Void` .

File: samples/client/petstore/swift4/promisekit/docs/FakeClassnameTags123API.md [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+
+
+# **testClassname**
+```swift
+    open class func testClassname( body: Client) -> Promise<Client>
+```
+
+To test class name in snake case
+
+To test class name in snake case
```
Comment (d-date) [inline samples/client/petstore/swift4/promisekit/docs/FakeClassnameTags123API.md]: I think some pattern has duplicated comments

File: samples/client/petstore/swift4/objcCompatible/docs/StoreAPI.md [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+### HTTP request headers
+
+ - **Content-Type**: Not defined
+ - **Accept**: application/json
+
+[[Back to top]](#) [[Back to API list]](../README.md#documentation-for-api-endpoints) [[Back to Model list]](../README.md#documentation-for-models) [[Back to README]](../README.md)
+
+# **getOrderById**
+```swift
+    open class func getOrderById(orderId: Int64, completion: @escaping ((_ data: Order?,_ error: Error?) -> Void))
```
Comment (d-date) [inline samples/client/petstore/swift4/objcCompatible/docs/StoreAPI.md]: should be 

- single bracket for closure
- [NIT] need space after comma


```swift
open class func getOrderById(orderId: Int64, completion: @escaping (_ data: Order?, _ error: Error?) -> Void)
```

**Comment by d-date** [comment]
I believe that all for review, but if you need to be more checked, let me know.

**Comment by wing328** [comment]
Thanks for the review 👍 

Let me fix those.

**Comment by wing328** [comment]
I should have fixed most if not all of the reviews.

About the duplicated description, it's due to both `summary` and `notes` set to the same description text.

**Comment by wing328** [comment]
Let's go with this one and we can always submit another PR to further improve the doc.

**Pull request merged**